### PR TITLE
Replace Gist links with embedded Gist

### LINKED ISSUE #494: Embed linked Gists
If we find a Gist link in a Markdown container, replace it with an embedded version of the Gist.

i.e.
```html
<a href="https://gist.github.com/jgierer12/06a8e3913ce4ae608c010b6f47efd257">https://gist.github.com/jgierer12/06a8e3913ce4ae608c010b6f47efd257</a>
```
would be replaced with (the result of)
```html
<script src="https://gist.github.com/jgierer12/06a8e3913ce4ae608c010b6f47efd257.js"></script>
```

Slightly related to #472/#473

### PR DESCRIPTION
**Pull request by donocode**
Replaces links with an embedded gist and also injects the stylesheet as a link. Only replaces links that are on their own line and in a comment body of a pr or issue.

Fixes #494

### CODE REVIEW

File: src/features/embed-gist-inline.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
+import select from 'select-dom';
+
+const gistRegex = /gist\.github\.com\/\w*?\/\w*/;
+const isGist = link => gistRegex.test(link.href);
```
Comment (sindresorhus) [inline src/features/embed-gist-inline.js]: Use https://github.com/sindresorhus/refined-github/blob/a8fd502d23084044c282bcdf56bac92666c2238f/src/libs/page-detect.js#L42

File: src/features/embed-gist-inline.js [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+const getJsonUrl = href => `${href}.json`.replace(/\.jso?n?\.json$/, '.json');
+const getGistData = href => fetch(getJsonUrl(href)).then(response => response.json());
+
+const createGistElement = gistData => {
+  const el = document.createElement('div');
+  const style = `<link rel="stylesheet" href="https://assets-cdn.github.com/assets/gist-embed-3cc724162479db25e452fdf621f2349adef3e742b53552c2a93f82d28156cb96.css" />`;
```
Comment (sindresorhus) [inline src/features/embed-gist-inline.js]: `" />` → `">`

File: src/features/embed-gist-inline.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+  el.innerHTML = style + gistData.div;
```
Comment (sindresorhus) [inline src/features/embed-gist-inline.js]: Do we need to do any sanitizing here?

File: src/features/embed-gist-inline.js [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+  return el;
+};
+
+export default async () => {
+  const gistLinks = select
+    .all('.js-comment-body p a[href^="https://gist.github.com"]:only-child')
+    .filter(isGist);
+
+  gistLinks.forEach(async link => {
```
Comment (sindresorhus) [inline src/features/embed-gist-inline.js]: Use a `for-of` loop

File: src/features/embed-gist-inline.js [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    try {
+      const gistData = await getGistData(link.href);
+      const gistEl = createGistElement(gistData);
+      const linkParent = link.parentNode;
+      linkParent.parentNode.replaceChild(gistEl, linkParent);
+    } catch (e) {}
```
Comment (sindresorhus) [inline src/features/embed-gist-inline.js]: `e` → `err`

File: src/features/embed-gist-inline.js [hunk 6]
```diff
@@ -0,0 +1,29 @@
+import select from 'select-dom';
+
+const gistRegex = /gist\.github\.com\/\w*?\/\w*/;
+const isGist = link => gistRegex.test(link.href);
+
+const getJsonUrl = href => `${href}.json`.replace(/\.jso?n?\.json$/, '.json');
+const getGistData = href => fetch(getJsonUrl(href)).then(response => response.json());
```
Comment (sindresorhus) [inline src/features/embed-gist-inline.js]: Use async/await.

File: src/features/embed-gist-inline.js [hunk 7]
```diff
@@ -0,0 +1,29 @@
+import select from 'select-dom';
+
+const gistRegex = /gist\.github\.com\/\w*?\/\w*/;
+const isGist = link => gistRegex.test(link.href);
+
+const getJsonUrl = href => `${href}.json`.replace(/\.jso?n?\.json$/, '.json');
```
Comment (sindresorhus) [inline src/features/embed-gist-inline.js]: I don't fully understand what `/\.jso?n?\.json$/` is doing. Like, why are we only replacing names that end with `.js` or `.json`? Can you show me an example URL? I think this would be clearer regardless `/\.(js|json)\.json$/`.

File: src/features/embed-gist-inline.js (hunk 1 above)
Comment (donocode) [reply src/features/embed-gist-inline.js]: This checks the current location. Do you mean to recreate the host or path check?

File: src/features/embed-gist-inline.js (hunk 7 above)
Comment (donocode) [reply src/features/embed-gist-inline.js]: It is for if the person links specifically the json or js version, so it won't create a link like https://gist.github.com/user/gist.js.json, it's probably unlikely for someone to be linking it but just in case

File: src/features/embed-gist-inline.js (hunk 3 above)
Comment (donocode) [reply src/features/embed-gist-inline.js]: The data from the gist is sanitised already, are we planning in case a gist response is somehow compromised?

File: src/features/embed-gist-inline.js (hunk 1 above)
Comment (sindresorhus) [reply src/features/embed-gist-inline.js]: Never mind. I misread the code

File: src/features/embed-gist-inline.js (hunk 3 above)
Comment (sindresorhus) [reply src/features/embed-gist-inline.js]: I don't know. I just asked in case anyone could think of a case. I guess it's fine then.

File: src/features/embed-gist-inline.js (hunk 7 above)
Comment (sindresorhus) [reply src/features/embed-gist-inline.js]: Ok, makes sense. I still would like to use `/\.(js|json)\.json$/` as it's more readable.

Can you also add a short comment above it to explain what it's doing.

File: src/features/embed-gist-inline.js (hunk 6 above)
Comment (sindresorhus) [reply src/features/embed-gist-inline.js]: ⬆️

File: src/features/embed-gist-inline.js [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+const getGistData = href => fetch(getJsonUrl(href)).then(response => response.json());
+
+const createGistElement = gistData => {
+	const el = (
```
Comment (sindresorhus) [inline src/features/embed-gist-inline.js]: No need to use a variable here. Just return it directly.

**Comment by sindresorhus** [comment]
Some more minor nitpick and it's good I think:

- https://github.com/sindresorhus/refined-github/pull/820#discussion_r153067128
- https://github.com/sindresorhus/refined-github/pull/820#discussion_r151592376
- https://github.com/sindresorhus/refined-github/pull/820#discussion_r153067185

**Comment by sindresorhus** [comment]
Why are we injecting the Gist instead of just loading the Gist embed in an iframe?

**Comment by donocode** [comment]
I'll get to this as soon as I can, I haven't forgotten about it!

File: src/features/embed-gist-inline.js (hunk 6 above)
Comment (bfred-it) [reply src/features/embed-gist-inline.js]: More context: https://github.com/sindresorhus/refined-github/pull/814

**Comment by donocode** [comment]
I took a look at an iframe solution, my main worry is the height as is always the case with iframes, it's difficult to set the height of the iframe to be correct for the height of the content. I can deal with this if required but do you think we are gaining from this?

**Comment by donocode** [comment]
It seems the linter does not like await in for of loops although (as far as I'm aware) that is correct usage.

File: src/features/embed-gist-inline.js [hunk 9]
```diff
@@ -1,30 +1,29 @@
-import { h } from 'dom-chef';
```
Comment (bfred-it) [inline src/features/embed-gist-inline.js]: Why was this removed? I think it’s required.

**Comment by bfred-it** [comment]
The linter complains because the loop stops at every await, which means that gists are loaded sequentially rather than in parallel. If you want to use await but load in parallel, extract the loop content into an async function and call it normally from inside the loop, don’t await the call.

**Comment by bfred-it** [comment]
@sindresorhus is there an iframe version? My understanding is that this is the preferred injection method. Their official embed code is a script tag with document.write

File: src/features/embed-gist-inline.js (hunk 9 above)
Comment (donocode) [reply src/features/embed-gist-inline.js]: Yes, good catch, my brain was still in iframe refactor mode when I deleted it

**Comment by bfred-it** [comment]
@sindresorhus this is the closest I got with an iframe, but it's sadly blocked by GH's CSP

```html
<iframe srcdoc="&lt;script src=&quot;https://gist.github.com/jgierer12/06a8e3913ce4ae608c010b6f47efd257.js&quot;&gt;&lt;/script&gt;">
```

But you're right, the current solution of injecting the code directly inline isn't working right because of conflicting styles.

Inline | Correct
-- | --
 | 

However the Shadow DOM seems to work correctly and is quite easy to implement with:

```js
link.attachShadow({mode: 'open'}).append(gistEl);
```

CSP also applies so it still requires the manual `fetch`

File: src/features/embed-gist-inline.js (hunk 7 above)
Comment (bfred-it) [reply src/features/embed-gist-inline.js]: I don't think `.js` and `.json` links should be embedded, they are technically not the same content (web page vs embed code)

**Comment by bfred-it** [comment]
This is mergeable for me, but:

- some gists may be long or have multiple files
- there's no clear indicator of what file belongs to what gist like for the regular github.com auto-embeds: 
    

- there's no loading indicator, the content just pops into place at some point in whatever loading order

**Comment by donocode** [comment]
Thanks for the help with this @bfred-it

File: source/features/embed-gist-inline.js [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+
+const createGistElement = gistData => (
+	<div>
+		<link rel="stylesheet" href={gistData.stylesheet} />
+		<div dangerouslySetInnerHTML={{__html: gistData.div}} />
+	</div>
+);
+
+async function embedGist(link) {
+	const response = await fetch(link.href + '.json');
```
Comment (sindresorhus) [inline source/features/embed-gist-inline.js]: Use template literal here.

**Comment by sindresorhus** [comment]
Looks great now :)

Last thing, can you mention it in the "new feature" highlight in the readme and link to a screenshot showing it in action? https://github.com/sindresorhus/refined-github#new-features

**Comment by donocode** [comment]
Add a cname:

https://gist.github.com/rotemtam/b4474f732674a9fd5884719d8c5bdf3e

**Comment by donocode** [comment]
embed-gist

**Comment by donocode** [comment]
That's all complete now @sindresorhus

**Comment by sindresorhus** [comment]
I opened a new issue for the things @bfred-it mentioned: https://github.com/sindresorhus/refined-github/issues/870
